### PR TITLE
Block Editor: Make sure unsaved changes popup appears in multi entity editing context

### PR DESCRIPTION
**Pull request by david-szabo97**
#### Changes proposed in this Pull Request

* Add a confirmation popup about unsaved changes to Back to dashboard button

#### Testing instructions

* Open site editor
* Make some changes
* Open navigation sidebar
* Click back until Back to dashboard available
* Click on Back to dashboard and make sure a confirmation alert pops up

Related to #

### CODE REVIEW

**Review by Addison-Stavlo — COMMENTED** [review]
The behavior is good, it works as expected.

I notice the alert author is a bit uglier than the post editors:

Screen Shot 2021-06-25 at 10 17 52 AM

vs. the post editor version:

Screen Shot 2021-06-25 at 10 18 29 AM

I wonder if we are able to use this [checkFormHandler](https://github.com/Automattic/wp-calypso/blob/fcc726e1d40c1c4d6f34b4d3499f237558dcf50d/client/lib/protect-form/index.tsx#L119) / [unsavedFormsMiddleware](https://github.com/Automattic/wp-calypso/blob/a6195b9b9dd74b5526c51f1cd0611be840bd63a0/client/boot/common.js#L197) in our site editor case somehow? 🤔  I think thats what the post editor is using.

**Comment by david-szabo97** [comment]
I think the author is different because you opened the site editor through wp-admin, but you opened the post editor through calypso. So in case of post editor wordpress.com/..... but in case of site editor you had yoursite.wordpress.com/.....

I don't think we can reuse that piece of code. What do you think about using the same alert text though? It would be consistent.

**Comment by Addison-Stavlo** [comment]
> What do you think about using the same alert text though? It would be consistent.

Makes sense!

File: apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ import { wordpress } from '@wordpress/icons';
 import { Component, useEffect, useState } from 'react';
 import tinymce from 'tinymce/tinymce';
 import debugFactory from 'debug';
+import { store as coreStore } from '@wordpress/core-data';
```
Comment (Addison-Stavlo) [inline apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: I wonder if its worth adding the extra dependency just for this? 🤔

**Review by Addison-Stavlo — APPROVED** [review]
> I think the author is different because you opened the site editor through wp-admin, but you opened the post editor through calypso. So in case of post editor wordpress.com/..... but in case of site editor you had yoursite.wordpress.com/.....

I opened them both through calypso.  The url that author is coming form is `WordPress.com/site-editor/{siteURL}`.  It would be nice if that could be consistent, but this confirmation window is working and is much better than not having one at all!

**Comment by david-szabo97** [comment]
@Addison-Stavlo Last commit should do it! Can you give it another round of testing? Thanks

File: apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js [hunk 2]
```diff
@@ -557,7 +557,9 @@ function handleCloseEditor( calypsoPort ) {
 			{
 				action: 'closeEditor',
 				payload: {
-					unsavedChanges: select( 'core/editor' ).isEditedPostDirty(),
+					unsavedChanges: editSitePackage
+						? select( 'core' ).__experimentalGetDirtyEntityRecords().length > 0
+						: select( 'core/editor' ).isEditedPostDirty(),
```
Comment (Addison-Stavlo) [inline apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: Im not sure checking for `editSitePackage` makes sense here as we are using a core selector.  Also, `__experimentalGetDirtyEntityRecords` could make sense in the post editor as well if they are using reusable blocks or site blocks, etc.  (I just noticed this bug extends to the post editor if the parent post isnt dirty but other things like reusable blocks are).

Perhaps we should do something like:
`
unsavedChanges: select( 'core' ).__experimenalGetDirtyEntityRecords?.().length > 0 || select( 'core/editor' ).isEditedPostDirty()
`

**Review by Addison-Stavlo — COMMENTED** [review]
This works well for the site editor, but I the post editor might benefit from `__experimentalGetDirtyEntityRecords` as well?

File: apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js (hunk 2 above)
Comment (david-szabo97) [reply apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: I was thinking about dropping `isEditedPostDirty` entirely, but it feels safer to have them both for back compact.

**Review by Addison-Stavlo — APPROVED** [review]
Looks and works great!

**Pull request merged**